samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemConsumer.java

```diff
@@ -330,6 +325,88 @@ public String getSystemName() {
     return systemName;
   }
 
+  @VisibleForTesting
+  class KafkaStartpointRegistrationHandler implements StartpointVisitor {
+
+    @Override
+    public void visit(SystemStreamPartition systemStreamPartition, StartpointSpecific startpointSpecific) {
+      TopicPartition topicPartition = toTopicPartition(systemStreamPartition);
+      long offsetInStartpoint = Long.parseLong(startpointSpecific.getSpecificOffset());
+      LOG.info("Updating the consumer fetch offsets of topic partition: {} to {}.", topicPartition, offsetInStartpoint);
+
+      // KafkaConsumer is not thread-safe.
+      synchronized (kafkaConsumer) {
+        kafkaConsumer.seek(topicPartition, offsetInStartpoint);
+      }
+
+      // add the partition to the proxy
+      proxy.addTopicPartition(systemStreamPartition, kafkaConsumer.position(topicPartition));
```

Comment:
Does `kafkaConsumer.position` need to be protected by the lock too?
Also, please check below for other usages of `kafkaConsumer` outside of the lock.

Reply:
Moved all the kafkaConsumer invocations into a lock.